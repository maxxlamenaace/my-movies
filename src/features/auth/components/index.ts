import AuthModal from './auth-modal/AuthModal';

export { AuthModal };
