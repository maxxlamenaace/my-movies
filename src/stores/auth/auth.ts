import create from 'zustand';
import { persist } from 'zustand/middleware';

type AuthState = {
  username?: string;
};

type AuthActions = {
  authenticate: (username: string) => void;
  isAuthenticated: () => boolean;
  logout: () => void;
};

export const useAuthStore = create(
  persist<AuthState & AuthActions>(
    (set, get) => ({
      authenticate: (username: string) =>
        set(() => ({
          username,
        })),
      isAuthenticated: () => {
        return get().username !== undefined;
      },
      logout: () =>
        set(() => ({
          username: undefined,
        })),
    }),
    { name: 'auth-storage' },
  ),
);
