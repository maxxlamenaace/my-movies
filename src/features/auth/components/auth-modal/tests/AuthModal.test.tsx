import React from 'react';

import { BrowserRouter } from 'react-router-dom';
import { render, act, fireEvent, cleanup, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { AuthModal } from '@/features/auth';

describe('AuthModal', () => {
  afterEach(() => {
    cleanup();
  });

  it('renders correctly', () => {
    const { baseElement } = render(<AuthModal isOpen={true} onSubmit={jest.fn()} />, {
      wrapper: BrowserRouter,
    });

    expect(baseElement).toMatchSnapshot();
  });

  describe('when user does not complete auth form', () => {
    it('renders with a disabled submit button', () => {
      const onSubmitMock = jest.fn();
      const { container } = render(<AuthModal isOpen={true} onSubmit={onSubmitMock} />, {
        wrapper: BrowserRouter,
      });

      const submitButton = container.querySelector('#auth-modal-submit-button');

      expect(submitButton).toBeDefined();

      act(() => {
        if (submitButton) {
          fireEvent.click(submitButton);
        }
      });

      waitFor(() => {
        expect(onSubmitMock).not.toHaveBeenCalled();
      });
    });
  });

  describe('when user completes the auth form and submit', () => {
    it('calls the onSubmit method with the auth form data', () => {
      const onSubmitMock = jest.fn();
      const { container } = render(<AuthModal isOpen={true} onSubmit={onSubmitMock} />, {
        wrapper: BrowserRouter,
      });

      const submitButton = container.querySelector('#auth-modal-submit-button');
      const textField = container.querySelector('#auth-modal-text-field');

      expect(submitButton).toBeDefined();
      expect(textField).toBeDefined();

      act(() => {
        if (submitButton && textField) {
          userEvent.type(textField, 'username');
          fireEvent.click(submitButton);
        }
      });

      waitFor(() => {
        expect(onSubmitMock).toHaveBeenCalledWith('username');
      });
    });
  });
});
