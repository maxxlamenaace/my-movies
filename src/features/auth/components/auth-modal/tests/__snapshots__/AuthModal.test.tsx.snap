// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthModal renders correctly 1`] = `
<body
  style="padding-right: 1024px; overflow: hidden;"
>
  <div
    aria-hidden="true"
  />
  <div
    class="MuiModal-root css-79ws1d-MuiModal-root"
    role="presentation"
  >
    <div
      aria-hidden="true"
      class="MuiBackdrop-root MuiModal-backdrop css-i9fmh8-MuiBackdrop-root-MuiModal-backdrop"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    />
    <div
      data-testid="sentinelStart"
      tabindex="0"
    />
    <div
      class="MuiBox-root css-w375ni"
      tabindex="-1"
    >
      <div
        class="MuiBox-root css-ga5x7p"
      >
        <div
          class="MuiBox-root css-13da791"
        >
          <div
            class="css-1p5q5e5-MuiStack-root"
          >
            <a
              href="/"
              id="logo-link"
              style="text-decoration: none;"
            >
              <p
                class="MuiTypography-root MuiTypography-body1 css-10bvdjj-MuiTypography-root"
                id="logo"
                style="color: rgba(0, 0, 0, 0.87);"
              >
                my
                <span
                  style="color: rgb(25, 118, 210);"
                >
                  Movies
                </span>
              </p>
            </a>
            <div
              class="css-1p5q5e5-MuiStack-root"
            >
              <p
                class="MuiTypography-root MuiTypography-body1 css-ndaqpv-MuiTypography-root"
              >
                Please enter your username ðŸ‘‡
              </p>
              <div
                class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
              >
                <div
                  class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
                >
                  <input
                    aria-invalid="false"
                    class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                    id="auth-modal-text-field"
                    placeholder="Username"
                    type="text"
                    value=""
                  />
                  <fieldset
                    aria-hidden="true"
                    class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                  >
                    <legend
                      class="css-ihdtdm"
                    >
                      <span
                        class="notranslate"
                      >
                        â€‹
                      </span>
                    </legend>
                  </fieldset>
                </div>
              </div>
              <button
                class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium Mui-disabled MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-sghohy-MuiButtonBase-root-MuiButton-root"
                disabled=""
                id="auth-modal-submit-button"
                tabindex="-1"
                type="submit"
              >
                Submit
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      data-testid="sentinelEnd"
      tabindex="0"
    />
  </div>
</body>
`;
